phylokit/balance.py

```diff
@@ -97,3 +108,59 @@ def b1_index(tree):
     return _b1_index(
         tree.postorder(), tree.left_child_array, tree.right_sib_array, tree.parent_array
     )
+
+
+@numba.njit(cache=True)
+def general_log(x, base):
+    """
+    Compute the logarithm of x in base `base`.
+
+    :param x: The number to compute the logarithm of.
+    :param base: The base of the logarithm.
+    :return: The logarithm of x in base `base`.
+    :rtype: float
+    """
+    return math.log(x) / math.log(base)
+
+
+@numba.njit(cache=True)
+def _b2_index(virtual_root, left_child, right_sib, base):
+    root = left_child[virtual_root]
+    stack = [(root, 1)]
+    total_proba = 0.0
+    while len(stack) > 0:
+        u, path_product = stack.pop()
+        if left_child[u] == -1:
+            total_proba -= path_product * general_log(path_product, base)
```

Comment:
We can probably just call ``math.log(path_product, base)`` here? I only did the ``general_log`` thing because C doesn't have this function.

Reply:
This is because the base argument is not implemented in Numba. 
In the future, we could probably move `general_log(x, base)` to `core.py` or something for any other potential usage?
```
There are 2 candidate implementations:
        - Of which 2 did not match due to:
        Type Restricted Function in function 'log': File: unknown: Line unknown.
          With argument(s): '(float64, float64)':
         No match for registered cases:
          * (int64,) -> float64
          * (uint64,) -> float64
          * (float32,) -> float32
          * (float64,) -> float64
```

Reply:
Aha, fair enough. All good then, let's leave it as it is. Let's not bother putting it into another file yet as we'd have to do more testing of the numerical corner cases.